### PR TITLE
AST parent pointer implementation

### PR DESCRIPTION
**Pull request by anweiss**
Implements #150.

Based on the work started by @SebastienGllmt in #134. Removes `parent` fields from the AST in favor of a rudimentary "arena" implementation which negates the need for AST manipulation. Introduces a `ParentVisitor` which houses the arena tree and implements the `Visitor`. Also introduces a `CDDLType` enum which holds all of the different AST nodes.

`ParentVisitor` and `CDDLType` can be used as follows to identify an AST node's parent:

```rust
let cddl_ast = cddl_from_str(r#"a = "myrule""#, true).unwrap();
let pv = ParentVisitor::new(&c).unwrap();
let rule = cddl_ast.rules.first().unwrap();

assert_eq!(
  CDDLType::from(rule).parent(&pv).unwrap(),
  &CDDLType::from(&cddl_ast)
);
```

You can also walk up the parent tree by recursively calling `parent()` on the returned `CDDLType` as follows:

```rust
let cddl_ast = cddl_from_str(r#"a = "myrule""#, true).unwrap();
let pv = ParentVisitor::new(&cddl_ast).unwrap();

if let Rule::Type { rule: tr, .. } = cddl_ast.rules.first().unwrap() {
  assert_eq!(
    CDDLType::from(tr).parent(&pv).unwrap().parent(&pv).unwrap(),
    &CDDLType::from(&cddl_ast)
   );
}
```

In the example above, the `TypeRule`'s parent's parent is the main `CDDL` struct itself.

`CDDLType::From` has also been implemented for all of the AST types.

@SebastienGllmt let me know if this API works for you.

### CODE REVIEW

**Comment by SebastienGllmt** [comment]
Slightly unfortunate this approach means you lose the type of the parent and have to use a broad `CDDLType` type as it will make pattern matching on the parent easier to get wrong. It might be better to bring back the original parent types from my original PR and put the casts somewhere in this library just to avoid downstream users having to think about it (ex: every node type could have a function that calls its own `parent` function and then performs the cast)

**Comment by anweiss** [comment]
Yea this was one of the trade offs vs having to manipulate the existing AST and storing the parents in the AST. But I agree that the specific concrete type of the parent should be returned. Will make some changes.

**Comment by anweiss** [comment]
@SebastienGllmt updated to abstract the `CDDLType` type and introduced a `Parent` trait which is implemented on the AST types. So you get this usage instead:

```rust
let cddl = cddl_from_str(r#"a = "myrule""#, true).unwrap();
let pv = ParentVisitor::new(&cddl).unwrap();
et rule = cddl.rules.first().unwrap();

assert_eq!(rule.parent(&pv).unwrap(), &cddl);
```

Parent trait definition:

```rust
pub trait Parent<'a, 'b: 'a, T> {
  fn parent(&'a self, parent_visitor: &'b ParentVisitor<'a, 'b>) -> Option<&T>;
}
```

File: src/validator/parent_visitor.rs [hunk 1]
```diff
@@ -0,0 +1,1030 @@
...
+      val,
+      parent: None,
+      children: vec![],
+    }
+  }
+}
+
+/// validator type
+// #[derive(Clone)]
+pub struct ParentVisitor<'a, 'b: 'a> {
```
Comment (SebastienGllmt) [inline src/validator/parent_visitor.rs:255]: I think you forgot to export this from `src/validator/mod.rs`

File: src/validator/parent_visitor.rs [hunk 2]
```diff
@@ -0,0 +1,1030 @@
...
+    }
+
+    self.arena_tree.arena[parent].children.push(child);
+
+    Ok(())
+  }
+}
+
+impl<'a, 'b: 'a> CDDLType<'a, 'b> {
+  fn parent(&self, visitor: &'b ParentVisitor<'a, 'b>) -> Option<&'b CDDLType<'a, 'b>> {
```
Comment (SebastienGllmt) [inline src/validator/parent_visitor.rs:291]: I assume this method is meant to be public?

File: src/validator/mod.rs [hunk 3]
```diff
@@ -5,6 +5,9 @@ pub mod cbor;
 /// JSON validation implementation
 pub mod json;
 
+/// parent visitor implementation
+mod parent_visitor;
```
[the remark this answers is not in the document]
Comment (anweiss) [reply src/validator/mod.rs:9]: Indeed ... there are some more commits incoming for this PR

File: src/validator/parent_visitor.rs (hunk 2 above)
Comment (SebastienGllmt) [reply src/validator/parent_visitor.rs:190]: How do I get the parent of a CDDLType representing an unknown AST element though?

**Comment by SebastienGllmt** [comment]
I think you forgot `RangeOp` / `CtlOp`

Also, CtlOp has a `ctr: &'a str` inside it, but I think this should be `Identifier`

File: src/validator/parent_visitor.rs (hunk 2 above)
Comment (anweiss) [reply src/validator/parent_visitor.rs:190]: Good point. Will make it public.

**Comment by anweiss** [comment]
@SebastienGllmt `RangeCtlOp` parent implemented. Moved the control operator tokens into a separate enum.

**Comment by anweiss** [comment]
@SebastienGllmt this is ready to merge

**Comment by SebastienGllmt** [comment]
LGTM. Integration worked as expected so far in our codebase 👍

**Pull request merged**